OpenRA.Game/FileSystem/BagFile.cs

```diff
@@ -0,0 +1,254 @@
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.Size + 36));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("WAVE"));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("fmt "));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(16));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)WavLoader.WaveType.Pcm));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)channels));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.SampleRate));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(2 * channels * entry.SampleRate));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)(2 * channels)));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)16));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("data"));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.Size));
+			}
+
+			if ((entry.Flags & 8) > 0)
+			{
+				// IMA ADPCM
+				var samplesPerChunk = (2 * (entry.Chunksize - 4)) + 1;
+				var bytesPerSec = (int)Math.Floor(((double)(2 * entry.Chunksize) / samplesPerChunk) * ((double)entry.SampleRate / 2));
+
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("RIFF"));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.Size + 52));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("WAVE"));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("fmt "));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(20));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)WavLoader.WaveType.ImaAdpcm));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)channels));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.SampleRate));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(bytesPerSec));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)entry.Chunksize));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)4));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)2));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes((short)samplesPerChunk));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("fact"));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(4));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.Chunksize * entry.Size));
+				waveHeaderMemoryStream.Write(Encoding.ASCII.GetBytes("data"));
+				waveHeaderMemoryStream.Write(BitConverter.GetBytes(entry.Size));
+			}
+
+			waveHeaderMemoryStream.Seek(0, SeekOrigin.Begin);
+
+			// Construct a merged stream
+			var mergedStream = new MergedStream(waveHeaderMemoryStream, s);
+			mergedStream.SetLength(waveHeaderMemoryStream.Length + entry.Size);
+
+			return mergedStream;
+		}
+
+		public bool Exists(string filename)
+		{
+			return audioIndex.ContainsKey(filename);
+		}
+
+		public IEnumerable<uint> ClassicHashes()
+		{
+			return Nothing;
+		}
+
+		public IEnumerable<uint> CrcHashes()
+		{
+			return Nothing;
```

Comment:
Returning nothing from both hash methods means that nothing will be cached and file lookups will always hit the slow path.  You should implement either this or ClassicHashes.